transport_nantes/topicblog/views.py

```diff
@@ -619,6 +625,105 @@ def form_valid(self, form):
+
+
+class TopicBlogSelfSendView(PermissionRequiredMixin, LoginRequiredMixin,
+                            SendableObjectMixin, BaseFormView):
+    """Allow authorised users to send a not published sendable TopicBlog
+    object to themselves."""
+    # Maybe a more specific permission would be better ?
+    permission_required = 'topicblog.tbe.may_send'
+    form_class = SendToSelfForm
+    send_record_class = None
+
+    def form_valid(self, form):
+        email = form.cleaned_data['email']
+        sent_object_id = form.cleaned_data['sent_object_id']
+        sent_object_class = form.cleaned_data['sent_object_class']
+        self.success_url = \
+            form.cleaned_data['redirect_url'] + "?email_sent=true"
+        self.context_appropriate_base_template = \
+            form.cleaned_data['context_appropriate_base_template']
+
+        self.send_email_to_self(email, sent_object_id, sent_object_class)
+        return super().form_valid(form)
+
+    def send_email_to_self(self, email: str, sent_object_id: int,
+                           sent_object_class: str):
+        """Send an email to the user with the given email address."""
+        object_class = apps.get_model('topicblog', sent_object_class)
+        object_to_send = object_class.objects.get(id=sent_object_id)
+        send_record = self.create_send_record(
+            sent_object_class, object_to_send, email)
+        if not send_record:
+            logger.info("No send record created, aborting the send.")
+            return None
+
+        context = self._set_email_context(
+            recipient=[email],
+            send_record_id=send_record.id,
+            tb_object=object_to_send
+        )
+        custom_email = self._create_email_object(
+            tb_object=object_to_send,
+            context=context,
+            recipient_list=[email],
+            send_record=send_record,
+        )
+        custom_email.send(fail_silently=False)
+
+    def create_send_record(
+        self,
+        sent_object_class: str,
+        object_to_send: Union[TopicBlogEmail, TopicBlogPress],
+        email: str) \
+            -> Union[Type[SendRecordTransactional], None]:
+        """Return the proper send record for the given class.
+
+        Keyword arguments:
+        sent_object_class -- the class of the object we are sending
+        object_to_send -- the object we are sending, a sendable Topicblog item
+        email -- the email address of the recipient
+
+        Returns:
+        The send record for the given class and email address.
+        Or if an error prevents the creation of a send record, None.
+        """
+        logger.info(f"Creating send record for {sent_object_class} ")
+        sent_object_class = sent_object_class.lower()
+
+        # Maybe we should put the SendRecord in the model ?
+        # i.e. topicblogemail.transactional_send_record = SendRecordTransactionalEmail # noqa
+        if sent_object_class == "topicblogemail":
```

Comment:
Use polymorphism, see below.  Or make `sent_object_class` be the actual name of the thing so that you can just call it.  Classes should never need to know a list of their possible derived classes.